File: packages/api/src/services/rateLimiter.ts
/**
 * Anthropic API Rate Limiter
 *
 * Implements sliding window rate limiting to prevent hitting API limits.
 * Tracks requests per minute (RPM) and tokens per minute (TPM).
 */

interface RateLimits {
  rpm: number; // Requests per minute
  inputTpm: number; // Input tokens per minute
  outputTpm: number; // Output tokens per minute
}

interface UsageEntry {
  timestamp: number;
  inputTokens: number;
  outputTokens: number;
}

// Anthropic rate limits by model tier
const MODEL_LIMITS: Record<string, RateLimits> = {
  haiku: { rpm: 50, inputTpm: 50000, outputTpm: 10000 },
  sonnet: { rpm: 50, inputTpm: 30000, outputTpm: 8000 },
  opus: { rpm: 50, inputTpm: 30000, outputTpm: 8000 },
};

// Buffer threshold - trigger rate limiting at this percentage of limit
const RATE_LIMIT_BUFFER = parseFloat(process.env.RATE_LIMIT_BUFFER || '0.8');
// Minimum delay between API calls (milliseconds)
const MIN_API_DELAY = parseFloat(process.env.MIN_API_DELAY || '0.5') * 1000;
// Debug logging
const RATE_LIMIT_DEBUG = process.env.RATE_LIMIT_DEBUG === 'true';

function debug(message: string, ...args: unknown[]): void {
  if (RATE_LIMIT_DEBUG) {
    console.log(`[RateLimiter] ${message}`, ...args);
  }
}

class AnthropicRateLimiter {
  private usageByTier: Map<string, UsageEntry[]> = new Map();
  private lastCallTime: number = 0;

  constructor() {
    // Initialize usage arrays for each tier
    for (const tier of Object.keys(MODEL_LIMITS)) {
      this.usageByTier.set(tier, []);
    }
  }

  /**
   * Get the model tier from a model name
   */
  getModelTier(model: string): string {
    const modelLower = model.toLowerCase();
    if (modelLower.includes('haiku')) return 'haiku';
    if (modelLower.includes('sonnet')) return 'sonnet';
    if (modelLower.includes('opus')) return 'opus';
    // Default to most restrictive tier
    return 'opus';
  }

  /**
   * Clean entries older than 60 seconds
   */
  private cleanOldEntries(tier: string): void {
    const entries = this.usageByTier.get(tier) || [];
    const cutoff = Date.now() - 60000; // 60 seconds ago
    const filtered = entries.filter((e) => e.timestamp > cutoff);
    this.usageByTier.set(tier, filtered);
  }

  /**
   * Get current usage in the sliding window
   */
  private getCurrentUsage(tier: string): {
    requests: number;
    inputTokens: number;
    outputTokens: number;
  } {
    this.cleanOldEntries(tier);
    const entries = this.usageByTier.get(tier) || [];

    return {
      requests: entries.length,
      inputTokens: entries.reduce((sum, e) => sum + e.inputTokens, 0),
      outputTokens: entries.reduce((sum, e) => sum + e.outputTokens, 0),
    };
  }

  /**
   * Calculate delay needed to stay under limits
   */
  private calculateDelay(
    tier: string,
    estimatedInputTokens: number,
    estimatedOutputTokens: number = 0
  ): number {
    const limits = MODEL_LIMITS[tier];
    if (!limits) return 0;

    const usage = this.getCurrentUsage(tier);
    const entries = this.usageByTier.get(tier) || [];

    // Calculate thresholds with buffer
    const rpmThreshold = limits.rpm * RATE_LIMIT_BUFFER;
    const inputTpmThreshold = limits.inputTpm * RATE_LIMIT_BUFFER;
    const outputTpmThreshold = limits.outputTpm * RATE_LIMIT_BUFFER;

    let maxDelay = 0;

    // Check RPM
    if (usage.requests >= rpmThreshold) {
      // Find oldest entry to determine when we can make another request
      const oldestEntry = entries.reduce(
        (min, e) => (e.timestamp < min ? e.timestamp : min),
        Date.now()
      );
      const rpmDelay = oldestEntry + 60000 - Date.now();
      if (rpmDelay > maxDelay) {
        maxDelay = rpmDelay;
        debug(
          `RPM limit: ${usage.requests}/${limits.rpm}, need delay: ${rpmDelay}ms`
        );
      }
    }

    // Check input TPM
    if (usage.inputTokens + estimatedInputTokens >= inputTpmThreshold) {
      // Find entries to remove to get under threshold
      let tokensToFree =
        usage.inputTokens + estimatedInputTokens - inputTpmThreshold;
      const sortedEntries = [...entries].sort(
        (a, b) => a.timestamp - b.timestamp
      );

      for (const entry of sortedEntries) {
        if (tokensToFree <= 0) break;
        const delay = entry.timestamp + 60000 - Date.now();
        if (delay > maxDelay) {
          maxDelay = delay;
          debug(
            `Input TPM limit: ${usage.inputTokens}/${limits.inputTpm}, need delay: ${delay}ms`
          );
        }
        tokensToFree -= entry.inputTokens;
      }
    }

    // Check output TPM
    if (usage.outputTokens + estimatedOutputTokens >= outputTpmThreshold) {
      let tokensToFree =
        usage.outputTokens + estimatedOutputTokens - outputTpmThreshold;
      const sortedEntries = [...entries].sort(
        (a, b) => a.timestamp - b.timestamp
      );

      for (const entry of sortedEntries) {
        if (tokensToFree <= 0) break;
        const delay = entry.timestamp + 60000 - Date.now();
        if (delay > maxDelay) {
          maxDelay = delay;
          debug(
            `Output TPM limit: ${usage.outputTokens}/${limits.outputTpm}, need delay: ${delay}ms`
          );
        }
        tokensToFree -= entry.outputTokens;
      }
    }

    // Enforce minimum delay between calls
    const timeSinceLastCall = Date.now() - this.lastCallTime;
    if (timeSinceLastCall < MIN_API_DELAY) {
      const minDelay = MIN_API_DELAY - timeSinceLastCall;
      if (minDelay > maxDelay) {
        maxDelay = minDelay;
      }
    }

    return Math.max(0, maxDelay);
  }

  /**
   * Wait for capacity before making an API call
   * Returns the number of milliseconds waited
   */
  async waitForCapacity(
    modelOrTier: string,
    estimatedInputTokens: number,
    estimatedOutputTokens: number = 0
  ): Promise<number> {
    const tier = MODEL_LIMITS[modelOrTier]
      ? modelOrTier
      : this.getModelTier(modelOrTier);
    const delay = this.calculateDelay(
      tier,
      estimatedInputTokens,
      estimatedOutputTokens
    );

    if (delay > 0) {
      console.log(
        `[RateLimiter] Waiting ${(delay / 1000).toFixed(1)}s for ${tier} capacity (estimated: ${estimatedInputTokens} input tokens)`
      );
      await new Promise((resolve) => setTimeout(resolve, delay));
    }

    this.lastCallTime = Date.now();
    return delay;
  }

  /**
   * Record actual token usage after an API call completes
   */
  recordUsage(
    modelOrTier: string,
    inputTokens: number,
    outputTokens: number
  ): void {
    const tier = MODEL_LIMITS[modelOrTier]
      ? modelOrTier
      : this.getModelTier(modelOrTier);
    const entries = this.usageByTier.get(tier) || [];

    entries.push({
      timestamp: Date.now(),
      inputTokens,
      outputTokens,
    });

    this.usageByTier.set(tier, entries);
    debug(
      `Recorded usage for ${tier}: ${inputTokens} input, ${outputTokens} output tokens`
    );
  }

  /**
   * Get current rate limit status for monitoring
   */
  getStatus(): Record<
    string,
    {
      requests: number;
      inputTokens: number;
      outputTokens: number;
      limits: RateLimits;
    }
  > {
    const status: Record<
      string,
      {
        requests: number;
        inputTokens: number;
        outputTokens: number;
        limits: RateLimits;
      }
    > = {};

    for (const tier of Object.keys(MODEL_LIMITS)) {
      const usage = this.getCurrentUsage(tier);
      status[tier] = {
        ...usage,
        limits: MODEL_LIMITS[tier],
      };
    }

    return status;
  }
}

// Export singleton instance
export const rateLimiter = new AnthropicRateLimiter();

// Export class for testing
export { AnthropicRateLimiter, MODEL_LIMITS };

File: packages/agents/src/monitoring/rate_limiter.py
"""
Anthropic API Rate Limiter

Implements sliding window rate limiting to prevent hitting API limits.
Tracks requests per minute (RPM) and tokens per minute (TPM).
Uses singleton pattern for global rate limiting across all agents.
"""

import os
import time
import threading
from collections import deque
from dataclasses import dataclass
from typing import Optional


@dataclass
class RateLimits:
    rpm: int  # Requests per minute
    input_tpm: int  # Input tokens per minute
    output_tpm: int  # Output tokens per minute


@dataclass
class UsageEntry:
    timestamp: float
    input_tokens: int
    output_tokens: int


# Anthropic rate limits by model tier
MODEL_LIMITS: dict[str, RateLimits] = {
    "haiku": RateLimits(rpm=50, input_tpm=50000, output_tpm=10000),
    "sonnet": RateLimits(rpm=50, input_tpm=30000, output_tpm=8000),
    "opus": RateLimits(rpm=50, input_tpm=30000, output_tpm=8000),
}

# Buffer threshold - trigger rate limiting at this percentage of limit
RATE_LIMIT_BUFFER = float(os.environ.get("RATE_LIMIT_BUFFER", "0.8"))
# Minimum delay between API calls (seconds)
MIN_API_DELAY = float(os.environ.get("MIN_API_DELAY", "0.5"))
# Debug logging
RATE_LIMIT_DEBUG = os.environ.get("RATE_LIMIT_DEBUG", "false").lower() == "true"


def debug(message: str) -> None:
    """Print debug message if debugging is enabled."""
    if RATE_LIMIT_DEBUG:
        print(f"[RateLimiter] {message}")


class AnthropicRateLimiter:
    """
    Thread-safe rate limiter for Anthropic API calls.
    Uses singleton pattern to ensure all agents share the same rate limit tracking.
    """

    _instance: Optional["AnthropicRateLimiter"] = None
    _lock = threading.Lock()

    def __new__(cls) -> "AnthropicRateLimiter":
        """Singleton pattern for global rate limiting."""
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = super().__new__(cls)
                    cls._instance._initialized = False
        return cls._instance

    def __init__(self):
        """Initialize the rate limiter (only runs once due to singleton)."""
        if self._initialized:
            return

        self._usage_by_tier: dict[str, deque[UsageEntry]] = {}
        self._last_call_time: float = 0
        self._tier_locks: dict[str, threading.Lock] = {}

        # Initialize usage deques and locks for each tier
        for tier in MODEL_LIMITS:
            self._usage_by_tier[tier] = deque()
            self._tier_locks[tier] = threading.Lock()

        self._initialized = True

    def get_model_tier(self, model: str) -> str:
        """Get the model tier from a model name."""
        model_lower = model.lower()
        if "haiku" in model_lower:
            return "haiku"
        if "sonnet" in model_lower:
            return "sonnet"
        if "opus" in model_lower:
            return "opus"
        # Default to most restrictive tier
        return "opus"

    def _clean_old_entries(self, tier: str) -> None:
        """Remove entries older than 60 seconds."""
        cutoff = time.time() - 60.0
        entries = self._usage_by_tier[tier]

        # Remove old entries from the left (deque is sorted by time)
        while entries and entries[0].timestamp < cutoff:
            entries.popleft()

    def _get_current_usage(self, tier: str) -> tuple[int, int, int]:
        """Get current usage in the sliding window: (requests, input_tokens, output_tokens)."""
        self._clean_old_entries(tier)
        entries = self._usage_by_tier[tier]

        requests = len(entries)
        input_tokens = sum(e.input_tokens for e in entries)
        output_tokens = sum(e.output_tokens for e in entries)

        return requests, input_tokens, output_tokens

    def _calculate_delay(
        self, tier: str, estimated_input_tokens: int, estimated_output_tokens: int = 0
    ) -> float:
        """Calculate delay needed to stay under limits (in seconds)."""
        limits = MODEL_LIMITS.get(tier)
        if not limits:
            return 0.0

        requests, input_tokens, output_tokens = self._get_current_usage(tier)
        entries = list(self._usage_by_tier[tier])

        # Calculate thresholds with buffer
        rpm_threshold = limits.rpm * RATE_LIMIT_BUFFER
        input_tpm_threshold = limits.input_tpm * RATE_LIMIT_BUFFER
        output_tpm_threshold = limits.output_tpm * RATE_LIMIT_BUFFER

        max_delay = 0.0
        now = time.time()

        # Check RPM
        if requests >= rpm_threshold:
            if entries:
                oldest_timestamp = min(e.timestamp for e in entries)
                rpm_delay = (oldest_timestamp + 60.0) - now
                if rpm_delay > max_delay:
                    max_delay = rpm_delay
                    debug(f"RPM limit: {requests}/{limits.rpm}, need delay: {rpm_delay:.1f}s")

        # Check input TPM
        if input_tokens + estimated_input_tokens >= input_tpm_threshold:
            tokens_to_free = (
                input_tokens + estimated_input_tokens - input_tpm_threshold
            )
            sorted_entries = sorted(entries, key=lambda e: e.timestamp)

            for entry in sorted_entries:
                if tokens_to_free <= 0:
                    break
                delay = (entry.timestamp + 60.0) - now
                if delay > max_delay:
                    max_delay = delay
                    debug(
                        f"Input TPM limit: {input_tokens}/{limits.input_tpm}, need delay: {delay:.1f}s"
                    )
                tokens_to_free -= entry.input_tokens

        # Check output TPM
        if output_tokens + estimated_output_tokens >= output_tpm_threshold:
            tokens_to_free = (
                output_tokens + estimated_output_tokens - output_tpm_threshold
            )
            sorted_entries = sorted(entries, key=lambda e: e.timestamp)

            for entry in sorted_entries:
                if tokens_to_free <= 0:
                    break
                delay = (entry.timestamp + 60.0) - now
                if delay > max_delay:
                    max_delay = delay
                    debug(
                        f"Output TPM limit: {output_tokens}/{limits.output_tpm}, need delay: {delay:.1f}s"
                    )
                tokens_to_free -= entry.output_tokens

        # Enforce minimum delay between calls
        time_since_last_call = now - self._last_call_time
        if time_since_last_call < MIN_API_DELAY:
            min_delay = MIN_API_DELAY - time_since_last_call
            if min_delay > max_delay:
                max_delay = min_delay

        return max(0.0, max_delay)

    def wait_for_capacity(
        self,
        model_or_tier: str,
        estimated_input_tokens: int,
        estimated_output_tokens: int = 0,
    ) -> float:
        """
        Block until capacity is available.

        Args:
            model_or_tier: Model name or tier (haiku/sonnet/opus)
            estimated_input_tokens: Estimated input tokens for the call
            estimated_output_tokens: Estimated output tokens (default 0)

        Returns:
            Number of seconds waited
        """
        tier = (
            model_or_tier
            if model_or_tier in MODEL_LIMITS
            else self.get_model_tier(model_or_tier)
        )

        with self._tier_locks[tier]:
            delay = self._calculate_delay(
                tier, estimated_input_tokens, estimated_output_tokens
            )

            if delay > 0:
                print(
                    f"[RateLimiter] Waiting {delay:.1f}s for {tier} capacity "
                    f"(estimated: {estimated_input_tokens} input tokens)"
                )
                time.sleep(delay)

            self._last_call_time = time.time()
            return delay

    def record_usage(
        self, model_or_tier: str, input_tokens: int, output_tokens: int
    ) -> None:
        """
        Record actual token usage after an API call completes.

        Args:
            model_or_tier: Model name or tier
            input_tokens: Actual input tokens used
            output_tokens: Actual output tokens used
        """
        tier = (
            model_or_tier
            if model_or_tier in MODEL_LIMITS
            else self.get_model_tier(model_or_tier)
        )

        with self._tier_locks[tier]:
            self._usage_by_tier[tier].append(
                UsageEntry(
                    timestamp=time.time(),
                    input_tokens=input_tokens,
                    output_tokens=output_tokens,
                )
            )
            debug(
                f"Recorded usage for {tier}: {input_tokens} input, {output_tokens} output tokens"
            )

    def get_status(self) -> dict[str, dict]:
        """Get current rate limit status for monitoring."""
        status = {}

        for tier, limits in MODEL_LIMITS.items():
            with self._tier_locks[tier]:
                requests, input_tokens, output_tokens = self._get_current_usage(tier)
                status[tier] = {
                    "requests": requests,
                    "input_tokens": input_tokens,
                    "output_tokens": output_tokens,
                    "limits": {
                        "rpm": limits.rpm,
                        "input_tpm": limits.input_tpm,
                        "output_tpm": limits.output_tpm,
                    },
                }

        return status


# Global singleton instance
rate_limiter = AnthropicRateLimiter()
